django/template/defaulttags.py

```diff
@@ -1564,3 +1567,171 @@ def do_with(parser, token):
+        self.inline = inline
+        self.nodelist = nodelist
+
+    def render(self, context):
+        return self.nodelist.render(context) if self.inline else ""
+
+
+class RenderPartialNode(Node):
+    def __init__(self, partial_name, partial_mapping):
+        # Defer lookup of nodelist to runtime.
+        self.partial_name = partial_name
+        self.partial_mapping = partial_mapping
+
+    def render(self, context):
+        return self.partial_mapping[self.partial_name].render(context)
+
+
+@register.tag(name="partialdef")
+def partialdef_func(parser, token):
+    """
+    Declare a partial that can be used later in the template.
+
+    Usage::
+
+        {% partialdef partial_name %}
+        Partial content goes here
+        {% endpartialdef %}
+
+    Stores the nodelist in the context under the key "partial_contents" and can
+    be retrieved using the {% partial %} tag.
+
+    The optional ``inline`` argument will render the contents of the partial
+    where it is defined.
+    """
+    tokens = token.split_contents()
+
+    # Check the number of tokens before trying to assign them via indexes.
+    if (tokens_len := len(tokens)) not in (2, 3):
+        name = token.contents.split()[0]
+        raise TemplateSyntaxError(f"{name} tag requires 2-3 arguments")
+
+    partial_name = tokens[1]
+    if tokens_len > 2:
+        inline = tokens[2]
+        if inline != "inline":
+            raise TemplateSyntaxError(
+                "The 'inline' argument does not have any parameters; "
+                "either use 'inline' or remove it completely."
+            )
+    else:
+        inline = False
+
+    # Parse the content until the end tag.
+    acceptable_endpartials = ("endpartialdef", f"endpartialdef {partial_name}")
+    nodelist = parser.parse(acceptable_endpartials)
+    endpartial = parser.next_token()
+    if endpartial.contents not in acceptable_endpartials:
+        parser.invalid_block_tag(endpartial, "endpartialdef", acceptable_endpartials)
+
+    # Store the partial nodelist in the parser.extra_data attribute.
+    parser.extra_data.setdefault("template-partials", {})
+    parser.extra_data["template-partials"][partial_name] = TemplateProxy(
+        nodelist, parser.origin, partial_name
+    )
+
+    return DefinePartialNode(partial_name, inline, nodelist)
+
+
+class SubDictionaryWrapper:
+    """
+    Wrap a parent dictionary, allowing deferred access to a sub-dictionary by key.
+    The parser.extra_data storage may not yet be populated when a partial node
+    is defined, so defer access until rendering.
+    """
+
+    def __init__(self, parent_dict, lookup_key):
+        self.parent_dict = parent_dict
+        self.lookup_key = lookup_key
+
+    def __getitem__(self, key):
+        try:
+            partials_content = self.parent_dict[self.lookup_key]
+        except KeyError:
+            raise TemplateSyntaxError(
+                f"No partials are defined. You are trying to access '{key}' partial"
```

Comment:
```suggestion
                f"Partial '{key}' could not be found because no partials are defined."
```

Reply:
There is a slight difference in these errors. The Key error you mentioned occurs when no partials are defined at all. The below KeyError occurs when some partials are defined, but the specific partial user is asking for is not defined. So I added that error to show the user that no partials are defined at all. 
I can change it if we want the same error. I understand that both error means the same. This just gives an extra hint that you have not defined any partials.

Reply:
Incorporated the change having the same error message makes testing easier. If we want, we can refactor if we want different error messages.